DA/Internal/Desugar.hs

```diff
@@ -150,6 +152,14 @@ toInterface = _toInterface
 
 class HasFromInterface t i where
   fromInterface : i -> Optional t
+  unsafeFromInterface : ContractId i -> i -> t
+
+class HasInterfaceView i v | i -> v
```

Comment:
Shouldn’t this have an `interfaceView : i -> v` method?

Reply:
No, we use the same trick as mkMethod/Method/HasMethod - we have an empty typeclass that forces the interface to view type dependency for a function call that we erase at LF conversion. This could be done using a type family, probably, but we don't support type families in LF as of right now.

Reply:
So how do I call the interface view in Daml? For the methods, I get a top-level definition per method but that doesn’t work here since then you get name collisions.

Reply:
You're right, I need to give this a bit more thought - I thought I was still desugaring view methods as methods, but I eliminate it from the method list, so that wasn't happening, and besides it wouldn't have worked anyways due to name clashes for views defined for each interface.

Reply:
I've opted to define a method `view` in `Desugar` as follows:

```
view : forall i t v. (Implements t i, HasInterfaceView i v) => t -> v
view = magic @"view" -- deleted by the compiler
```

This gets erased and replaced with `EViewInterface` during LFConversion. We still use mkInterfaceView/InterfaceView/HasInterfaceView so that viewtypes can be defined by interfaces and then carried across module boundaries.

I *think* this is the best way to carry the type dependency across module boundaries.

Reply:
How does that interact with monomorphization? `EViewInterface` requires that you pass in concrete template/interface typecons. Usually that works because we define these types of methods as part of a typeclass so within the instance we are working with concrete arguments. If you define this outside, it seems like you somehow have to make this polymorphic and that doesn’t seem to match the primitive you're suggesting in https://github.com/digital-asset/daml/pull/14486/files#diff-7d491e6cd5a6cd6421f36491226e588057a509d4ca3ada1c277ee9b0748fd572

Reply:
Switching to @remyhaemmerle-da’s suggested option of only having the interface typecon in there won’t remove the need for that afaict.

Reply:
@remyhaemmerle-da's suggestion of keeping only the interface typecon _and_ expecting the argument to be monomorphically of type `$interface` simplifies things a bit since we can ignore the `$template` type, but that still leaves the interface type like @cocreature said.

For that, I think it should be enough add a method to `HasInterfaceView`, with a slightly different type to allow its use on `$template` values:

```haskell
class HasInterfaceView i v | i -> v where
  _view : forall t. Implements t i => t -> v
  -- The compiler will complain about `i` being ambiguous, since it isn't mentioned in 
  -- the RHS of the type.
  -- We silence this with {-# LANGUAGE AllowAmbiguousTypes #-}. All this means is that users
  -- will be required to provide the interface type with a type applications, which is
  -- already the case for functions such as `exerciseByKey`

-- I define `view` separately to show that it has the same type as yours and to 
-- give it a top level signature. It's also needed to reorder the type variables, 
-- otherwise `v` would come before `t`.
view : forall i t v. (Implements t i, HasInterfaceView i v) => t -> v
view = _view @i
```
The RdrHsSyn-generated instances for this will look like

```haskell
instance HasInterfaceView $i $v where
  _view = GHC.Types.primitiveInterfaceView (DA.Internal.Desugar._toInterface @_ @$i t)
  -- Note that `$i` will always correspond to a concrete interface TyCon. 
```

`primitiveInterfaceView` would be a new dummy function defined next to [`primitiveInterface`](https://github.com/digital-asset/daml/blob/main/compiler/damlc/daml-prim-src/GHC/Types.daml#L152) which you would pick up during lf conversion and translate into `EViewInterface`, the code should be very similar to that for [`primitiveInterface`](https://github.com/digital-asset/daml/blob/main/compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs#L1239-L1249). The trick is that the generated calls to `primitiveInterfaceView` are always applied to the result of `_toInterface`, so it doesn't need to deal with template-typed values. Then, since `_toInterface` is always applied to a concrete interface, the argument type will be available in `go` and `convertType` will give you the corresponding tycon.

Reply:
In my iteration yesterday/this morning I opted for

```
class HasInterfaceView i v | i -> v where
  _view : i -> v

view : forall i t v. (Implements t i, HasInterfaceView i v) => t -> v
view t = _view (_toInterface @t @i t)
```

And defined `_view = primitive @"EViewInterface"` in all instances, and adapted `LFConversion/Primitives.hs`. This mirrors how we implement other typeclass-oriented primitives.

wdyt?

Reply:
Aah that's even better, since `convertPrim` for `EViewInterface` will only look at the type where it’s used, as opposed to `convertExpr.go` for `primitiveInterfaceView`, which needs to look at the GHC-Core type arguments.

The important part is that `_view` has to be defined per-interface, since that’s where the type gets monomorphized into an interface tycon. Since your code does that too, I think we're good to go.